core/apple/src/SourcesApple.kt

```diff
@@ -0,0 +1,164 @@
+            { false }
+        }
+    }
+
+    private var status = NSStreamStatusNotOpen
+    private var error: NSError? = null
+        set(value) {
+            status = NSStreamStatusError
+            field = value
+            postEvent(NSStreamEventErrorOccurred)
+            source.close()
+        }
+
+    override fun streamStatus() = if (isClosed()) NSStreamStatusClosed else status
+
+    override fun streamError() = error
+
+    override fun open() {
+        if (status == NSStreamStatusNotOpen) {
+            status = NSStreamStatusOpen
+            postEvent(NSStreamEventOpenCompleted)
+            checkBytes()
+        }
+    }
+
+    override fun close() {
+        if (status == NSStreamStatusError) return
+        status = NSStreamStatusClosed
+        source.close()
+        runLoop = null
+        runLoopModes = listOf()
+    }
+
+    override fun read(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+        try {
+            if (isClosed()) throw IOException("Underlying source is closed.")
+            if (status != NSStreamStatusOpen && status != NSStreamStatusAtEnd) {
+                return -1
+            }
+            if (source.exhausted()) {
+                status = NSStreamStatusAtEnd
+                return 0
+            }
+            if (buffer == null) return -1
+
+            status = NSStreamStatusReading
+            val toRead = minOf(maxLength.toInt(), source.buffer.size).toInt()
+            val read = source.buffer.readAtMostTo(buffer, toRead).convert<NSInteger>()
+            status = NSStreamStatusOpen
+            checkBytes()
+            return read
+        } catch (e: Exception) {
+            error = e.toNSError()
+            return -1
+        }
+    }
+
+    override fun getBuffer(buffer: CPointer<CPointerVar<uint8_tVar>>?, length: CPointer<NSUIntegerVar>?) = false
+
+    override fun hasBytesAvailable() = !source.exhausted()
+
+    override fun propertyForKey(key: NSStreamPropertyKey): Any? = null
+
+    override fun setProperty(property: Any?, forKey: NSStreamPropertyKey) = false
+
+    private var _delegate = WeakReference<NSStreamDelegateProtocol>(this)
+    private var runLoop: NSRunLoop? = null
+    private var runLoopModes = listOf<NSRunLoopMode>()
+
+    private fun postEvent(event: NSStreamEvent) {
+        val runLoop = runLoop ?: return
+        runLoop.performInModes(runLoopModes) {
+            if (runLoop == this.runLoop) {
+                delegate?.stream(this, event)
+            }
+        }
+    }
+
+    private fun checkBytes(sendEndEvent: Boolean = true) {
+        val runLoop = runLoop ?: return
+        runLoop.performInModes(runLoopModes) {
+            if (runLoop != this.runLoop) return@performInModes
+            try {
+                if (source.exhausted()) {
+                    if (sendEndEvent) delegate?.stream(this, NSStreamEventEndEncountered)
+                    val timer = NSTimer.timerWithTimeInterval(0.1, false) {
```

Comment:
Perhaps it'd make sense to use a back off algorithm here instead of a fixed poll interval, since the frequency of byte availability from a given `Source` is unknown and would vary with different use cases.

Alternatively, the poll interval could be made configurable, possibly with a key value passed to `setProperty()`.

Reply:
I was thinking there may potentially be other `Source` implementations that could have bytes become available in the future, like a socket protocol. I'd assume the implementation would ensure thread safety either internally or by scheduling itself on the same run loop as the consumer. This is correct though, that my test using `Buffer` is not thread safe. It does make sense `NSStreamEventEndEncountered` could be considered a terminal event to close the stream. I can just remove the polling.

Reply:
> I was thinking there may potentially be other Source implementations that could have bytes become available in the future, like a socket protocol.

That's something one can actually anticipate, but the Source's documentation explicitly states that `exhausted` means that the source is actually exhausted and no data is expected once `Source::exhausted` returned true. Source's underlying source is considered exhausted when a read form it returned `-1` and that particular value should only be returned if there will be no more data in the future (and if we expect to receive more bytes then the read should return `0`).
In case with sockets, it's up to an implementor how it'll work internally, but corresponding source's `readAtMostTo` should return `-1` only when a connection was closed, I guess.
So it's better to expect that particular `RawSource`/`RawSink` implementations correctly implement and obey the contract, instead of trying to handle their misbehavior.
`Buffer` is the outlier here, I can say nothing to defend its behavior. :)